python/ray/data/_internal/execution/interfaces.py

```diff
@@ -445,6 +453,17 @@ def incremental_resource_usage(self) -> ExecutionResources:
         """
         return ExecutionResources()
 
+    def notify_resource_usage(
+        self, input_queue_size: int, under_resource_limits: bool
+    ) -> None:
+        """Called periodically by the executor.
+
+        Args:
+            input_queue_size: The number of inputs queued outside this operator.
+            under_resource_limits: Whether this operator is under resource limits.
```

Comment:
just for my understanding, how come the operator needs to be told if it's under resource limits instead of knowing this information itself?

Reply:
This is since the limit is across all the operators, so only the executor knows the full picture.